freeciv-web/src/main/webapp/javascript/improvement.js

```diff
@@ -19,6 +19,9 @@
 
 var improvements = {};
 
+// incomplete list of (well defined) building IDs - populate as needed
+const B_AIRPORT = 0;
+const B_PALACE = 21;
```

Comment:
That's 23 with civ2civ3 ruleset.

Reply:
Dynamic? Good catch - back to the drawing board.

FWIW, I find our PR back-and-fourth very useful in understanding "edge" (sometimes not so edge...) which need to be checked for working. Thanks again, for your help and diligence.

I expect it has come up, but is there a discussion going on anywhere around automating a test suite and how to set up a framework for testing various edge cases - something more targeted than just running an autogame for N turns?

Reply:
Not that I knew it depended on ruleset. I found out when looking for the source of that 21.

We had a similar problem with unit orders that depend on technologies. As a "temporary" fix, the ids were changed to a search by name (https://github.com/freeciv/freeciv-web/commit/1e3ec23ade0b6038ca7a1149381a22aa36b4b789, https://github.com/freeciv/freeciv-web/commit/60ad3e448d7eb84eeffd46ff91025a3140f1ca1d). The same search should work here with the `improvements` list.

Re test, I'm not aware of any discussion, neither automatic nor manual, not even a checklist. I guess it could be done, with a large sample of crafted savegames and casper/selenium/whatever scripts to check values and the results of different actions. It would also be a hard, thankless job.

Reply:
I suspect the dynamic `EXTRA_*` (858cc4c6) variables I encountered in (ba71af03) are related. Any recommendations on how to get up-to speed on Effects is welcome - I see [effects.*](https://github.com/freeciv/freeciv/blob/27a777a32ef0faa0bcca819c10f2fab4dcf15d51/common/effects.c) and other refernces, but no prose description in the [HACKING](https://github.com/freeciv/freeciv/blob/master/doc/HACKING) file - I find discussions of intent and descriptions of design decisions very useful guides before diving into what I like to call "code archaeology". I'll eventually end up looking in forums or mailing lists, but if you know of discussions, it would be great to mention them in #208.

Reply:
fixed (see commit message).